controllers/hyperparametertuningjob/hyperparametertuningjob_controller.go

```diff
@@ -18,505 +18,339 @@ package hyperparametertuningjob
+		}
+	}
+
+	// Spawn training jobs regardless of the status
+	ctx.HPOTrainingJobSpawner.SpawnMissingTrainingJobs(ctx, *ctx.TuningJob)
+	if err = r.addBestTrainingJobToStatus(ctx); err != nil {
+		return r.updateStatusAndReturnError(ctx, ReconcilingTuningJobStatus, errors.Wrap(err, "Unable to add best training job to status"))
+	}
 
-		log.Info("Job is stopping, nothing to do")
-		if err := r.updateJobStatus(ctx, hpojobv1.HyperparameterTuningJobStatus{
-			HyperParameterTuningJobStatus:        string(ctx.SageMakerDescription.HyperParameterTuningJobStatus),
-			LastCheckTime:                        Now(),
-			SageMakerHyperParameterTuningJobName: *ctx.Job.Spec.HyperParameterTuningJobName,
-			TrainingJobStatusCounters:            newTrainingJobStatusCountersFromDescription(ctx.SageMakerDescription),
-		}); err != nil {
-			return RequeueIfError(err)
+	switch ctx.TuningJobDescription.HyperParameterTuningJobStatus {
+	case sagemaker.HyperParameterTuningJobStatusInProgress:
+		if controllers.HasDeletionTimestamp(ctx.TuningJob.ObjectMeta) {
+			// Request to stop the job. If SageMaker returns a 404 then the job has already been deleted.
+			if _, err := ctx.SageMakerClient.StopHyperParameterTuningJob(ctx, ctx.TuningJobName); err != nil && !clientwrapper.IsStopHyperParameterTuningJob404Error(err) {
+				return r.updateStatusAndReturnError(ctx, ReconcilingTuningJobStatus, errors.Wrap(err, "Unable to delete hyperparameter tuning job"))
+			}
+			// Describe the new state of the job
+			if ctx.TuningJobDescription, err = ctx.SageMakerClient.DescribeHyperParameterTuningJob(ctx, ctx.TuningJobName); err != nil {
+				return r.updateStatusAndReturnError(ctx, ReconcilingTuningJobStatus, errors.Wrap(err, "Unable to describe SageMaker hyperparameter tuning job"))
+			}
 		}
-		return RequeueAfterInterval(r.PollInterval, nil)
-	case sagemaker.HyperParameterTuningJobStatusCompleted, sagemaker.HyperParameterTuningJobStatusFailed, sagemaker.HyperParameterTuningJobStatusStopped:
-		log.Info("Job is in terminal state. Deleting spawned TrainingJobs and removing finalizer.")
-
-		// Delete all spawned TrainingJobs, retry if failure.
-		err := ctx.HpoTrainingJobSpawner.DeleteSpawnedTrainingJobs(ctx, ctx.Job)
-		if err != nil {
-			log.Info("Not all associated TrainingJobs jobs were deleted, will retry", "error", err)
-			return RequeueAfterInterval(r.PollInterval, nil)
+
+	case sagemaker.HyperParameterTuningJobStatusStopped, sagemaker.HyperParameterTuningJobStatusFailed, sagemaker.HyperParameterTuningJobStatusCompleted:
+		if controllers.HasDeletionTimestamp(ctx.TuningJob.ObjectMeta) {
+			return r.cleanupAndRemoveFinalizer(ctx)
 		}
 
-		return r.removeFinalizerAndUpdate(ctx)
+	case sagemaker.HyperParameterTuningJobStatusStopping:
+		break
+
 	default:
-		log.Info("Job is in unknown status")
-		return NoRequeue()
+		unknownStateError := errors.New(fmt.Sprintf("Unknown Tuning Job Status: %s", ctx.TuningJobDescription.HyperParameterTuningJobStatus))
```

Comment:
We should use `fmt.Errorf(...)` instead of `errors.New(fmt.Sprintf(...))`. I know you were just copying from old TrainingJob, this was a thing that I learned between writing TJ and Hosting.